### PR TITLE
Add CPUThreads resource and Progress Meters

### PR DESCRIPTION
**Pull request by OkonSamuel**
- [x] Add `acceleration = CPUThreads` Computational Resource (closes issue #27 and #15)
- [x] Add progress meters for all computational resources.(`CPUThreads`, `CPUProcesses` and `CPU1`) (closes issue #14)
- [x]  issue an informative warning for poor choices of combining Computational Resourceses. (closes issue #39)

For this to work well the following [PR](https://github.com/alan-turing-institute/MLJBase.jl/pull/283) at [MLJBase](https://github.com/alan-turing-institute/MLJBase.jl/pull/283) needs to be merged first

### CODE REVIEW

**Comment by ablaom** [comment]
cc @OkonSamuel https://github.com/JuliaRegistries/General/pull/14033

**Comment by OkonSamuel** [comment]
@ablaom . A quick review would be appreciated

File: src/tuned_models.jl [hunk 1]
```diff
@@ -278,27 +291,134 @@ function event(metamodel,
...
+                         history,
+                         state,
+                         CPU1())
+   end
+    n_metamodels = length(metamodels)
+    n_threads = Threads.nthreads()
+    
+    
+    ret = Array{Any, 1}(undef, n_metamodels)
+    verbosity < 1 || (p = Progress(n_metamodels,
```
Comment (ablaom) [inline src/tuned_models.jl:379]: Be nice if we could avoid this abstract `Any` typing of `ret`, but I'm sure you've thought about that. I doubt it's a real performance issue, except possibly when dealing with very small data and huge numbers of models (hyper parameters) being evaluated.

File: src/tuned_models.jl [hunk 2]
```diff
@@ -278,27 +291,134 @@ function event(metamodel,
...
+                   machine(Resampler(model= resampling_machines[1].model.model,
+                      resampling    = resampling_machines[1].model.resampling,
+                      measure       = resampling_machines[1].model.measure,
+                      weights       = resampling_machines[1].model.weights,
+                      operation     = resampling_machines[1].model.operation,
+                      check_measure = resampling_machines[1].model.check_measure,
+                      repeats       = resampling_machines[1].model.repeats,
+                      acceleration  = resampling_machines[1].model.acceleration),
+                      resampling_machines[1].args...)
+            end
```
Comment (ablaom) [inline src/tuned_models.jl:404]: Nice. You avoid deep copy here. I had to think very carefully about the `model.model` but I see that this get's *replaced* and not just mutated in the `event` call below.

File: test/tuned_models.jl [hunk 3]
```diff
@@ -77,18 +78,30 @@ end
...
+    tm = TunedModel(model=first(r), tuning=Explicit(),
+                    range=r, resampling=CV(nfolds=2),
+                    measures=[rms, l1], acceleration=CPUProcesses(),
+                    acceleration_resampling=accel)
+    fitresult, meta_state, report = fit(tm, 0, X, y);
+    history, _, state = meta_state;
+    results4 = map(event -> last(event).measurement[1], history)
+    @test results4 ≈ results
+end
+
```
Comment (ablaom) [inline test/tuned_models.jl:104]: Nice. You've tested all combinations of the two acceleration params and they are giving the same results.

File: src/tuned_models.jl (hunk 1 above)
Comment (OkonSamuel) [reply src/tuned_models.jl:379]: `ret = Vector{Tuple{M,Any}(undef, n_metamodels)` is the best i can do since the type of `result` isn't constrained.

Comment (ablaom) [reply src/tuned_models.jl:379]: No that's fine.

**Comment by OkonSamuel** [comment]
> rocket This is a stellar contribution. I can see you have spent many hours getting under the hood of MLJTuning! I am most grateful.
> 
> I did notice one thing, which is that the progress bars do not appear in the case of learning curves in the two parallel modes. In fact, increasing verbosity to 4 did not expose any logging at all. Most curious.
> 
> I would be happy for you to pull this when ready, and tag a new patch release. Or ping me and I will do it.
> 
> Again many thanks.

I checked it seems the verbosity was fixed at `-1` in the code

**Comment by OkonSamuel** [comment]
@ablaom. Its finally ready and everything show well on my PC (You can also check to see if everything is okay by you). Made some slight changes (switched from `floor` to `ceil` otherwise two or more tasks running on the same thread may interfere with each other shared mutable state.) 
Just merge [this](https://github.com/alan-turing-institute/MLJBase.jl/pull/289), tag a new release so i can rebuild this PR again.

**Comment by ablaom** [comment]
Okay working on a release. Just want to slip in a bug fix.

**Comment by ablaom** [comment]
https://github.com/JuliaRegistries/General/pull/14235

**Comment by OkonSamuel** [comment]
@ablaom. It all done.

**Comment by OkonSamuel** [comment]
changed `pmap` to `@distributed`. Noticed that for one worker process nesting `pmap` in `@distributed` (which occurs in `learning_curve` method) hangs.

**Comment by ablaom** [comment]
> changed pmap to @distributed. Noticed that for one worker process nesting pmap in @distributed (which occurs in learning_curve method) hangs.

Good catch. Again many thanks. Just merge this yourself when you are happy and I'll tag a new patch release if you don't want to do that yourself.

**Comment by OkonSamuel** [comment]
> Just merge this yourself

@ablaom. I don't have merge rights. Anyway it ready to be merged

**Comment by ablaom** [comment]
> I don't have merge rights. Anyway it ready to be merged

😳 Oh, sorry. I guess I didn't put you onto this repo yet. Just invited you for future.

Let me know if there are any other MLJ repos you want access to and I will add you.

**Pull request merged**